### DISCUSSION TITLE
Welcome to qvapay-python Discussions!

### DISCUSSION THREAD

**Discussion by CarlosLugones** [ragnarok22/qvapay-python#4] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by CarlosLugones** [comment]
#3 Async operations. How to do it?

There is `aioqvapay` based on this lib, is already doing this, we should join efforts.

https://github.com/leynier/aioqvapay

Opening discussion about this.

**Comment by codeshard** [comment]
As as suggested before, the work of @leynier is awesome but rely in the use of [pydantic](https://pydantic-docs.helpmanual.io/), lib that I absolutely love, but, trying to keep the qvapay deps as shallow as possible, I used dataclasses and a wrapper for serialization and field aliases and other perks called dataclasses-json.

Now, for async/sync support, may I suggest refactor the actual library and make the Client initialization rely on one attr that defines whether use or not async/sync

Example:
```python
c = Client(app_id, app_secret, version, async=True)
```
 
With this flag we could load the httpx async o sync client and the functions (few of them) on demand.

Plz @lugodev make an issue with this.
Also, the may I suggest the project structure of @leynier 's

**Reply by CarlosLugones** [reply]
I agree, we shoud use `async` attribute on `Client` class to handle async/sync cases.

IDK about pydanting, I have to document myself about that lib.

Let's then keep the deps as shallow as possible, that should be the right approach.

Issue #3 is open to handle this case.

I'm checking @leynier project structure. Still thinking we should join efforts.

**Reply by leynier** [reply]
So the proposal is?

1. Use `dataclasses` without `pydantic`.
2. Use the **"same"** method and that it behaves *asynchronous* or *synchronous* depending on the parameter with which the `Client` class was initialized.

What concerns me from point 2 is to get type annotation correctly so that IDEs interpret cases correctly. If this cannot be done, I would prefer to use different methods getting better IDE auto-completion and being more explicit (explicit better than implicit)

**Reply by leynier** [reply]
It would also be good other things like publishing on Pypi using GitHub Actions, Coverage report, testing, and other things that I added to [aioqvapay](https://github.com/leynier/aioqvapay), which I would like to add if possible in this project.

**Reply by CarlosLugones** [reply]
We might use different methods. But, when endpoints number increases, there will be a lot of methods.

What you think @codeshard?

Let's consider this.

**Reply by CarlosLugones** [reply]
@leynier that would be great! Feel free to add GitHub actions or anything you consider to improve the lib.

Would you need any permissions to do that?

**Reply by leynier** [reply]
I also think it would be nice to completely separate version 1, version 2, etc into different paths. Because right now the changes are minimal, but there can be big changes between one version or another, and keeping those changes in the same Client would be entangled.

**Reply by leynier** [reply]
> We might use different methods. But, when endpoints number increases, there will be a lot of methods.
> 
> What you think @codeshard?
> 
> Let's consider this.

It will be kept in 2 * N methods where N is the number of endpoints.

**Reply by leynier** [reply]
> @leynier that would be great! Feel free to add GitHub actions or anything you consider to improve the lib.
> 
> Would you need any permissions to do that?

I would only need to set some SECRETS in the repository, that I can tell you what they are for you to do.

**Reply by leynier** [reply]
@codeshard can you show us a summary of your idea to unify the async and sync behavior in the "same" method?

**Reply by CarlosLugones** [reply]
> I also think it would be nice to completely separate version 1, version 2, etc into different paths. Because right now the changes are minimal, but there can be big changes between one version or another, and keeping those changes in the same Client would be entangled.

Agree, we should do that.

**Reply by CarlosLugones** [reply]
@leynier accept invitation to receive permissions.

**Reply by leynier** [reply]
> @leynier accept invitation to receive permissions.

To handle the SECRETS, apparently you must have TOP-LEVEL permissions, so don't worry, if I do something that needs them, I'll tell you to do it yourself.

**Reply by CarlosLugones** [reply]
There might be a way, I'll check that.

**Reply by leynier** [reply]
> @codeshard can you show us a summary of your idea to unify the async and sync behavior in the "same" method?

So, let's wait for @codeshard's answer to decide which route we will take.

**Reply by CarlosLugones** [reply]
New cover!

![image](https://user-images.githubusercontent.com/18733370/131413769-11015c39-3fde-4838-b1f4-5e34419f8b55.png)

**Reply by codeshard** [reply]
> @codeshard can you show us a summary of your idea to unify the async and sync behavior in the "same" method?

My suggestion is, as I briefly explain in the PR, pass a "flag" to the client, and in this way load the async / sync functions, keeping the same codebase, @leynier you did something similar in your lib, so, basically, if the flag is True, we load the async client, otherwise, the sync client. 

Btw, there's and stage API http://stage.qvapay.com/docs/. I've tested all end endpoints and all of them works, plz review the models and check every attr. cc @leynier 

Ok, let's make a checklist:

- [x] different paths for API versioning
- [x] async support
- [x] publishing on Pypi using GitHub Actions, Coverage report, testing
- [ ] guarantee backward compatibility with major python versions


Btw, @leynier mi disclaimer with pydantic ( I really really love pydantic, btw), in this case will be overkill and over-engineered lib, QvaPay API so far, only have a couple of endpoints, and the schema generation, serialization and data validation could be achieved with built-in libs, btw, I missed `AnyHttpUrl`

**Reply by leynier** [reply]
> > @codeshard can you show us a summary of your idea to unify the async and sync behavior in the "same" method?
> 
> My suggestion is, as I briefly explain in the PR, pass a "flag" to the client, and in this way load the async / sync functions, keeping the same codebase, @leynier you did something similar in your lib, so, basically, if the flag is True, we load the async client, otherwise, the sync client.
> 
> Btw, there's and stage API http://stage.qvapay.com/docs/. I've tested all end endpoints and all of them works, plz review the models and check every attr. cc @leynier
> 
> Ok, let's make a checklist:
> 
> * [ ]  different paths for API versioning
> * [ ]  async support
> * [ ]  publishing on Pypi using GitHub Actions, Coverage report, testing
> * [ ]  guarantee backward compatibility with major python versions
> 
> Btw, @leynier mi disclaimer with pydantic ( I really really love pydantic, btw), in this case will be overkill and over-engineered lib, QvaPay API so far, only have a couple of endpoints, and the schema generation, serialization and data validation could be achieved with built-in libs, btw, I missed `AnyHttpUrl`

The problem is the following: How to make that depending on a variable of class `A`, the method `f` of class `A` is **asynchronous** or **synchronous** maintaining the type annotation. I think if this was possible, `httpx` would have done it. The way `httpx`  is to create two client classes, one synchronous and one asynchronous. The way I used in `aioqvapay` was a single client class, but with an asynchronous alternative for each method called `<methodname>_async`, similar to how Entity Framework does in ASP.Net Core.

Which do you prefer? Is there another alternative?

**Reply by CarlosLugones** [reply]
Good morning folks! I'm solving a problem here, after I finish with that, I'm joining the debate. Thanks for your efforts!

**Reply by codeshard** [reply]
> So the proposal is?
> 
> 1. Use `dataclasses` without `pydantic`.
> 2. Use the **"same"** method and that it behaves _asynchronous_ or _synchronous_ depending on the parameter with which the `Client` class was initialized.
> 
> What concerns me from point 2 is to get type annotation correctly so that IDEs interpret cases correctly. If this cannot be done, I would prefer to use different methods getting better IDE auto-completion and being more explicit (explicit better than implicit)

Agreed, using the same method will be a total mess

**Reply by leynier** [reply]
Super!!! I will be preparing a pull request with:

- `async`/`await` support
- A `Client` with a specific path for each version of API
- Testing with code coverage 100%
- Deploy to PyPi and GitHub Releases base on tags with `v*.*.*` format
- Check Pythonic style with `flake8`, `black` and `isort`
- `class` for obtain auth params (`api_id` and `app_secret`) from `.env` file or from environment

**Reply by codeshard** [reply]
@leynier Highlander!!

**Reply by leynier** [reply]
Guys, I've been going through the pros and cons of using pydantic or dataclasses + dataclasses-json and I think pydantic is better.  The point that pydantic has more internal dependencies I do not consider true, I checked the dependencies of each alternative and pydantic has less internal dependencies (for example dataclasses-json depends on marshmallow and others more) in addition to using pydantic to make the linter (in dataclasses -json the from_dict the linter does not recognize it) in addition to having pydantic type checking the tests can easily validate that the models are correct.  So I think it is better to use pydantic.  What do you think?

**Reply by codeshard** [reply]
Pydantic is way better! That's a fact. I suggest using dataclasses 'cause pydantic is a huge, as 3rd party package, and for a lib that will consume a few endpoints is overkill. For sure I'll embrace pydantic usage, mostly, I use it everyday 😄

**Reply by leynier** [reply]
@lugodev @codeshard good morning guys, PR => https://github.com/lugodev/qvapay-python/pull/7

**Reply by CarlosLugones** [reply]
Hi guys, sorry for the delay. I was (am) sick, possible covid. Let's see #7. Thanks for your great work @leynier. :tada:

**Reply by CarlosLugones** [reply]
New version is almost ready guys, but there is some CI test failing. Can you please check that?

Solving that, we can release v0.0.4.

@codeshard @leynier

https://github.com/lugodev/qvapay-python/pull/7#issuecomment-913111734
